File: day10/day10.py
#!/usr/bin/env python3


import sys

data = open("input.txt" if len(sys.argv) == 1 else sys.argv[1]).read().splitlines()


def check(line):
    stack = []
    corrupted = {")": 3, "]": 57, "}": 1197, ">": 25137}
    completed = {")": 1, "]": 2, "}": 3, ">": 4}

    for c in line:
        if c == "<":
            stack.append(">")
        elif c == "(":
            stack.append(")")
        elif c == "[":
            stack.append("]")
        elif c == "{":
            stack.append("}")
        else:
            d = stack.pop()
            if c != d:
                return corrupted[c], 0

    score = 0
    while stack:
        score = score * 5 + completed[stack.pop()]

    return 0, score


part1 = 0
part2 = []
for line in data:
    corrupted, completed = check(line)
    part1 += corrupted
    if completed != 0:
        part2.append(completed)
print(part1)
part2 = sorted(part2)
print(part2[len(part2) // 2])

File: src/year2021/day10/day10.rs
//! [Day 10: Syntax Scoring](https://adventofcode.com/2021/day/10)

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (u64, u64) {
    let mut part1 = 0;
    let mut part2 = vec![];

    for line in data.lines() {
        let (corrupted, completed) = check(line);

        part1 += corrupted;
        if completed != 0 {
            part2.push(completed);
        }
    }

    // part2.sort_by(|a, b| a.cmp(b));
    part2.sort_unstable();

    (part1, part2[part2.len() / 2])
}

fn check(line: &str) -> (u64, u64) {
    let mut stack = vec![];

    for c in line.chars() {
        match c {
            '(' => stack.push(')'),
            '[' => stack.push(']'),
            '{' => stack.push('}'),
            '<' => stack.push('>'),
            _ => {
                let d = stack.pop().unwrap();
                if c != d {
                    match c {
                        ')' => return (3, 0),
                        ']' => return (57, 0),
                        '}' => return (1197, 0),
                        '>' => return (25137, 0),
                        _ => return (0, 0),
                    }
                }
            }
        }
    }

    let mut score = 0u64;
    while let Some(d) = stack.pop() {
        match d {
            ')' => score = score * 5 + 1,
            ']' => score = score * 5 + 2,
            '}' => score = score * 5 + 3,
            '>' => score = score * 5 + 4,
            _ => score *= 5,
        }
    }

    (0, score)
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test_solve() {
        assert_eq!(solve(TEST_INPUT), (26397, 288957));
    }
}
